File: ruby/10.rb
def sieve(limit)
  current = 2
  numbers = 0.upto(limit).to_a
  numbers[1] = 0
  while (current < limit)
    (2 * current .. limit).step(current).each { |i| numbers[i] = 0 }
    begin
      current += 1
    end while current <= limit && numbers[current] == 0
  end
  numbers.select { |n| n != 0 }
end

result = sieve(2_000_000).inject(&:+)

puts result
raise Error unless result == 142913828922

File: scala/10.scala
def sieve(limit: Int) = {
  var current = 2
  var numbers = (0 to limit).toArray
  numbers(1) = 0
  while (current <= limit) {
    for (i <- 2 * current to limit by current) numbers(i) = 0
    do { current += 1 } while (current <= limit && numbers(current) == 0)
  }
  numbers.filter(_ != 0).toList
}

println(sieve(2000000).map(_.toLong).reduceLeft(_ + _))
